# HG changeset patch
# User alta88@fixall.com
# Date 1701465917 -3600
# Parent  77e1d5b77cafe83c78519d72c546258034c11910
Feature: Add Wide Thread and Stacked layout views to 3pane.

Wide thread layout: thread tree across the top, and folder tree and message browser on the bottom.
Stacked layout: folder tree and thread tree on the left, and message browser on the right.

Authored by Alta88 and contributed to the Betterbird Project.
See https://github.com/Betterbird/thunderbird-patches/blob/main/LICENSE for license details.
All uses require attribution to the Author.

diff --git a/mail/base/content/about3Pane.js b/mail/base/content/about3Pane.js
--- a/mail/base/content/about3Pane.js
+++ b/mail/base/content/about3Pane.js
@@ -145,17 +145,17 @@ window.addEventListener("unload", () => 
 });
 
 var paneLayout = {
   init() {
     this.folderPaneSplitter = document.getElementById("folderPaneSplitter");
     this.messagePaneSplitter = document.getElementById("messagePaneSplitter");
 
     for (let [splitter, properties, storeID] of [
-      [this.folderPaneSplitter, ["width"], "folderPaneBox"],
+      [this.folderPaneSplitter, ["height", "width"], "folderPaneBox"],
       [this.messagePaneSplitter, ["height", "width"], "messagepaneboxwrapper"],
     ]) {
       for (let property of properties) {
         let value = Services.xulStore.getValue(XULSTORE_URL, storeID, property);
         if (value) {
           splitter[property] = value;
         }
       }
@@ -193,36 +193,57 @@ var paneLayout = {
       (name, oldValue, newValue) => this.setLayout(newValue)
     );
     this.setLayout(this.layoutPreference);
     threadPane.updateThreadView(
       Services.xulStore.getValue(XULSTORE_URL, "threadPane", "view")
     );
   },
 
+  LAYOUTS: ["standard", "wide", "vertical", "widethread", "stacked"],
+
+  CONFIG_MAP: {
+    standard: {
+      className: "layout-classic",
+      folderPaneSplitterDir: "horizontal",
+      messagePaneSplitterDir: "vertical",
+    },
+    wide: {
+      className: "layout-wide",
+      folderPaneSplitterDir: "horizontal",
+      messagePaneSplitterDir: "vertical",
+    },
+    vertical: {
+      className: "layout-vertical",
+      folderPaneSplitterDir: "horizontal",
+      messagePaneSplitterDir: "vertical",
+    },
+    widethread: {
+      className: "layout-widethread",
+      folderPaneSplitterDir: "horizontal",
+      messagePaneSplitterDir: "vertical",
+    },
+    stacked: {
+      className: "layout-stacked",
+      folderPaneSplitterDir: "vertical",
+      messagePaneSplitterDir: "horizontal",
+    },
+  },
+
   setLayout(preference) {
-    document.body.classList.remove(
-      "layout-classic",
-      "layout-vertical",
-      "layout-wide"
-    );
-    switch (preference) {
-      case 1:
-        document.body.classList.add("layout-wide");
-        this.messagePaneSplitter.resizeDirection = "vertical";
-        break;
-      case 2:
-        document.body.classList.add("layout-vertical");
-        this.messagePaneSplitter.resizeDirection = "horizontal";
-        break;
-      default:
-        document.body.classList.add("layout-classic");
-        this.messagePaneSplitter.resizeDirection = "vertical";
-        break;
-    }
+    this.LAYOUTS.forEach(layout => {
+      document.body.classList.remove(this.CONFIG_MAP[layout].className);
+    });
+
+    let layoutConfig = this.CONFIG_MAP[this.LAYOUTS[preference]];
+    document.body.classList.add(layoutConfig.className);
+    this.folderPaneSplitter.resizeDirection =
+      layoutConfig.folderPaneSplitterDir;
+    this.messagePaneSplitter.resizeDirection =
+      layoutConfig.messagePaneSplitterDir;
   },
 
   get accountCentralVisible() {
     return document.body.classList.contains("account-central");
   },
   get folderPaneVisible() {
     return !this.folderPaneSplitter.isCollapsed;
   },
@@ -2578,20 +2599,25 @@ var folderPane = {
 
       MailE10SUtils.loadURI(
         accountCentralBrowser,
         `chrome://messenger/content/msgAccountCentral.xhtml?folderURI=${encodeURIComponent(
           gFolder.URI
         )}`
       );
       document.body.classList.add("account-central");
+      paneLayout.folderPaneSplitter.resizeDirection = "horizontal";
       accountCentralBrowser.hidden = false;
     } else {
       document.title = `${gFolder.name} - ${gFolder.server.prettyName}`;
       document.body.classList.remove("account-central");
+      let layoutConfig =
+        paneLayout.CONFIG_MAP[paneLayout.LAYOUTS[paneLayout.layoutPreference]];
+      paneLayout.folderPaneSplitter.resizeDirection =
+        layoutConfig.folderPaneSplitterDir;
       accountCentralBrowser.hidden = true;
 
       quickFilterBar.activeElement = null;
       threadPane.restoreColumns();
 
       gViewWrapper = new DBViewWrapper(dbViewWrapperListener);
 
       threadPane.scrollToNewMessage =
@@ -6953,16 +6979,22 @@ commandController.registerCallback("cmd_
   Services.prefs.setIntPref("mail.pane_config.dynamic", 0)
 );
 commandController.registerCallback("cmd_viewWideMailLayout", () =>
   Services.prefs.setIntPref("mail.pane_config.dynamic", 1)
 );
 commandController.registerCallback("cmd_viewVerticalMailLayout", () =>
   Services.prefs.setIntPref("mail.pane_config.dynamic", 2)
 );
+commandController.registerCallback("cmd_viewWideThreadMailLayout", () =>
+  Services.prefs.setIntPref("mail.pane_config.dynamic", 3)
+);
+commandController.registerCallback("cmd_viewStackedMailLayout", () =>
+  Services.prefs.setIntPref("mail.pane_config.dynamic", 4)
+);
 commandController.registerCallback(
   "cmd_toggleThreadPaneHeader",
   () => threadPaneHeader.toggleThreadPaneHeader(),
   () => gFolder && !gFolder.isServer
 );
 commandController.registerCallback(
   "cmd_toggleFolderPane",
   () => paneLayout.folderPaneSplitter.toggleCollapsed(),
diff --git a/mail/base/content/mainCommandSet.inc.xhtml b/mail/base/content/mainCommandSet.inc.xhtml
--- a/mail/base/content/mainCommandSet.inc.xhtml
+++ b/mail/base/content/mainCommandSet.inc.xhtml
@@ -42,16 +42,18 @@
      <command id="cmd_viewPageSource" oncommand="goDoCommand('cmd_viewPageSource')" disabled="true"/>
      <command id="cmd_setFolderCharset" oncommand="goDoCommand('cmd_setFolderCharset')" />
 
      <command id="cmd_expandAllThreads" oncommand="goDoCommand('cmd_expandAllThreads')" disabled="true"/>
      <command id="cmd_collapseAllThreads" oncommand="goDoCommand('cmd_collapseAllThreads')" disabled="true"/>
      <command id="cmd_viewClassicMailLayout" oncommand="goDoCommand('cmd_viewClassicMailLayout')" disabled="true"/>
      <command id="cmd_viewWideMailLayout" oncommand="goDoCommand('cmd_viewWideMailLayout')" disabled="true"/>
      <command id="cmd_viewVerticalMailLayout" oncommand="goDoCommand('cmd_viewVerticalMailLayout')" disabled="true"/>
+     <command id="cmd_viewWideThreadMailLayout" oncommand="goDoCommand('cmd_viewWideThreadMailLayout')" disabled="true"/>
+     <command id="cmd_viewStackedMailLayout" oncommand="goDoCommand('cmd_viewStackedMailLayout')" disabled="true"/>
      <command id="cmd_toggleFolderPane" oncommand="goDoCommand('cmd_toggleFolderPane')" disabled="true"/>
      <command id="cmd_toggleThreadPaneHeader" oncommand="goDoCommand('cmd_toggleThreadPaneHeader')" disabled="true"/>
      <command id="cmd_toggleMessagePane" oncommand="goDoCommand('cmd_toggleMessagePane')" disabled="true"/>
      <command id="cmd_viewAllMsgs" oncommand="goDoCommand('cmd_viewAllMsgs')" disabled="true"/>
      <command id="cmd_viewUnreadMsgs" oncommand="goDoCommand('cmd_viewUnreadMsgs')" disabled="true"/>
      <command id="cmd_viewThreadsWithUnread" oncommand="goDoCommand('cmd_viewThreadsWithUnread')" disabled="true"/>
      <command id="cmd_viewWatchedThreadsWithUnread" oncommand="goDoCommand('cmd_viewWatchedThreadsWithUnread')" disabled="true"/>
      <command id="cmd_viewIgnoredThreads" oncommand="goDoCommand('cmd_viewIgnoredThreads')" disabled="true"/>
diff --git a/mail/base/content/messenger-menubar.inc.xhtml b/mail/base/content/messenger-menubar.inc.xhtml
--- a/mail/base/content/messenger-menubar.inc.xhtml
+++ b/mail/base/content/messenger-menubar.inc.xhtml
@@ -361,16 +361,26 @@
     <menu id="menu_MessagePaneLayout" label="&messagePaneLayoutStyle.label;" accesskey="&messagePaneLayoutStyle.accesskey;">
       <menupopup id="view_layout_popup" onpopupshowing="InitViewLayoutStyleMenu(event)">
         <menuitem id="messagePaneClassic" type="radio" label="&messagePaneClassic.label;" name="viewlayoutgroup"
                   accesskey="&messagePaneClassic.accesskey;" command="cmd_viewClassicMailLayout"/>
         <menuitem id="messagePaneWide" type="radio" label="&messagePaneWide.label;" name="viewlayoutgroup"
                   accesskey="&messagePaneWide.accesskey;" command="cmd_viewWideMailLayout"/>
         <menuitem id="messagePaneVertical" type="radio" label="&messagePaneVertical.label;" name="viewlayoutgroup"
                   accesskey="&messagePaneVertical.accesskey;" command="cmd_viewVerticalMailLayout"/>
+        <menuitem id="messagePaneWideThread"
+                  type="radio"
+                  data-l10n-id="menu-view-layout-widethread"
+                  name="viewlayoutgroup"
+                  command="cmd_viewWideThreadMailLayout"/>
+        <menuitem id="messagePaneStacked"
+                  type="radio"
+                  data-l10n-id="menu-view-layout-stacked"
+                  name="viewlayoutgroup"
+                  command="cmd_viewStackedMailLayout"/>
         <menuseparator id="viewMenuAfterPaneVerticalSeparator"/>
         <menuitem id="multilineAllFolders"
                   type="checkbox"
                   data-l10n-id="menu-multi-line-all-folders"
                   oncommand="ToogleMultilineAll(false);"/>
         <menuseparator id="viewMenuBeforeShowFolderPaneSeparator"/>
         <menuitem id="menu_showFolderPane" type="checkbox" label="&showFolderPaneCmd.label;"
                   accesskey="&showFolderPaneCmd.accesskey;" command="cmd_toggleFolderPane"/>
diff --git a/mail/components/customizableui/content/panelUI.inc.xhtml b/mail/components/customizableui/content/panelUI.inc.xhtml
--- a/mail/components/customizableui/content/panelUI.inc.xhtml
+++ b/mail/components/customizableui/content/panelUI.inc.xhtml
@@ -349,16 +349,28 @@
                        name="viewlayoutgroup"
                        command="cmd_viewWideMailLayout"/>
         <toolbarbutton id="appmenu_messagePaneVertical"
                        class="subviewbutton subviewbutton-iconic"
                        type="radio"
                        label="&messagePaneVertical.label;"
                        name="viewlayoutgroup"
                        command="cmd_viewVerticalMailLayout"/>
+        <toolbarbutton id="appmenu_messagePaneWideThread"
+                       class="subviewbutton subviewbutton-iconic"
+                       type="radio"
+                       data-l10n-id="menu-view-layout-widethread"
+                       name="viewlayoutgroup"
+                       command="cmd_viewWideThreadMailLayout"/>
+        <toolbarbutton id="appmenu_messagePaneStacked"
+                       class="subviewbutton subviewbutton-iconic"
+                       type="radio"
+                       data-l10n-id="menu-view-layout-stacked"
+                       name="viewlayoutgroup"
+                       command="cmd_viewStackedMailLayout"/>
         <toolbarseparator id="appmenu_viewMenuAfterPaneVerticalSeparator"/>
         <toolbarbutton id="appmenu_multilineAllFolders"
                        class="subviewbutton subviewbutton-iconic"
                        type="checkbox"
                        data-l10n-id="menu-multi-line-all-folders"
                        oncommand="ToogleMultilineAll(true);"/>
         <toolbarseparator id="appmenu_viewMenuBeforeShowFolderPaneSeparator"/>
         <toolbarbutton id="appmenu_showFolderPane"
diff --git a/mail/components/extensions/parent/ext-mailTabs.js b/mail/components/extensions/parent/ext-mailTabs.js
--- a/mail/components/extensions/parent/ext-mailTabs.js
+++ b/mail/components/extensions/parent/ext-mailTabs.js
@@ -13,17 +13,17 @@ XPCOMUtils.defineLazyModuleGetters(this,
 
 XPCOMUtils.defineLazyPreferenceGetter(
   this,
   "gDynamicPaneConfig",
   "mail.pane_config.dynamic",
   0
 );
 
-const LAYOUTS = ["standard", "wide", "vertical"];
+const LAYOUTS = ["standard", "wide", "vertical", "widethread", "stacked"];
 // From nsIMsgDBView.idl
 const SORT_TYPE_MAP = new Map(
   Object.keys(Ci.nsMsgViewSortType).map(key => {
     // Change "byFoo" to "foo".
     let shortKey = key[2].toLowerCase() + key.substring(3);
     return [Ci.nsMsgViewSortType[key], shortKey];
   })
 );
diff --git a/mail/components/extensions/schemas/mailTabs.json b/mail/components/extensions/schemas/mailTabs.json
--- a/mail/components/extensions/schemas/mailTabs.json
+++ b/mail/components/extensions/schemas/mailTabs.json
@@ -50,17 +50,17 @@
           },
           "viewType": {
             "type": "string",
             "optional": true,
             "enum": ["ungrouped", "groupedByThread", "groupedBySortType"]
           },
           "layout": {
             "type": "string",
-            "enum": ["standard", "wide", "vertical"]
+            "enum": ["standard", "wide", "vertical", "widethread", "stacked"]
           },
           "folderPaneVisible": {
             "type": "boolean",
             "optional": true
           },
           "messagePaneVisible": {
             "type": "boolean",
             "optional": true
diff --git a/mail/locales/en-US/messenger/menubar.ftl b/mail/locales/en-US/messenger/menubar.ftl
--- a/mail/locales/en-US/messenger/menubar.ftl
+++ b/mail/locales/en-US/messenger/menubar.ftl
@@ -112,16 +112,24 @@ menu-view-folders-toggle-header =
     .label = Folder Pane Header
     .accesskey = F
 
 menu-multi-line-all-folders =
     .label = Multi-line View On All Folders
 
 ## View / Layout
 
+menu-view-layout-widethread =
+    .label = Wide Thread View
+    .accesskey = T
+
+menu-view-layout-stacked =
+    .label = Stacked View
+    .accesskey = S
+
 menu-view-toggle-thread-pane-header =
     .label = Message List Header
     .accesskey = H
 
 menu-font-size-label =
     .label = Font Size
     .accesskey = o
 
diff --git a/mail/themes/shared/mail/about3Pane.css b/mail/themes/shared/mail/about3Pane.css
--- a/mail/themes/shared/mail/about3Pane.css
+++ b/mail/themes/shared/mail/about3Pane.css
@@ -100,16 +100,17 @@ html {
 
 body {
   height: 100%;
   margin: 0;
   text-shadow: none;
 
   display: grid;
   --folderPaneSplitter-width: 18em;
+  --folderPaneSplitter-height: 18em;
   --messagePaneSplitter-width: 54em;
   --messagePaneSplitter-height: 36em;
 }
 
 /* Different layouts */
 
 #folderPane {
   grid-area: folders;
@@ -190,16 +191,31 @@ body.layout-vertical {
 /* Wide layout: folder tree and thread tree on the top, and message browser on the bottom. */
 body.layout-wide {
   grid-template: "folders folderPaneSplitter threads" minmax(auto, 1fr)
                  "messagePaneSplitter messagePaneSplitter messagePaneSplitter" min-content
                  "message message message" minmax(auto, var(--messagePaneSplitter-height))
                  / minmax(auto, var(--folderPaneSplitter-width)) min-content minmax(auto, 1fr);
 }
 
+/* Wide thread layout: thread tree across the top, and folder tree and message browser on the bottom. */
+body.layout-widethread {
+  grid-template: "threads threads threads" minmax(auto, 1fr)
+                 "messagePaneSplitter messagePaneSplitter messagePaneSplitter" min-content
+                 "folders folderPaneSplitter message" minmax(auto, var(--messagePaneSplitter-height))
+                 / minmax(auto, var(--folderPaneSplitter-width)) min-content minmax(auto, 1fr);
+}
+
+/* Stacked layout: folder tree and thread tree on the left, and message browser on the right. */
+body.layout-stacked {
+  grid-template: "folders messagePaneSplitter message" minmax(auto, var(--folderPaneSplitter-height))
+                 "folderPaneSplitter messagePaneSplitter message" min-content
+                 "threads messagePaneSplitter message" minmax(auto, 1fr)
+                 / minmax(auto, 1fr) min-content minmax( var(--messagePaneSplitter-width), 1fr);
+}
 /* If Account Central is shown, it overrides the layout setting. */
 body.account-central {
   grid-template: "folders folderPaneSplitter account-central" auto
                  / minmax(auto, var(--folderPaneSplitter-width)) min-content minmax(auto, 1fr);
 }
 
 body.account-central :is(
   #threadPane,
